processing/src/main/java/org/apache/druid/query/expression/NestedDataExpressions.java

```diff
@@ -442,38 +442,76 @@ public String name()
     public Expr apply(List<Expr> args)
     {
       final List<NestedPathPart> parts = getArg1JsonPathPartsFromLiteral(name(), args);
-      class JsonValueExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
-      {
-        public JsonValueExpr(List<Expr> args)
-        {
-          super(name(), args);
+      if (args.size() == 3 && args.get(2).isLiteral()) {
+        final ExpressionType castTo = ExpressionType.fromString((String) args.get(2).getLiteralValue());
+        if (castTo == null) {
+          throw new IAE("Invalid output type: [%s]", args.get(2).getLiteralValue());
         }
-
-        @Override
-        public ExprEval eval(ObjectBinding bindings)
+        class CastJsonValueExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
         {
-          ExprEval input = args.get(0).eval(bindings);
-          return ExprEval.bestEffortOf(
-              NestedPathFinder.findLiteral(unwrap(input), parts)
-          );
-        }
+          public CastJsonValueExpr(List<Expr> args)
+          {
+            super(name(), args);
+          }
 
-        @Override
-        public Expr visit(Shuttle shuttle)
-        {
-          List<Expr> newArgs = args.stream().map(x -> x.visit(shuttle)).collect(Collectors.toList());
-          return shuttle.visit(new JsonValueExpr(newArgs));
-        }
+          @Override
+          public ExprEval eval(ObjectBinding bindings)
+          {
+            ExprEval input = args.get(0).eval(bindings);
+            return ExprEval.bestEffortOf(
+                NestedPathFinder.findLiteral(unwrap(input), parts)
+            ).castTo(castTo);
+          }
 
-        @Nullable
-        @Override
-        public ExpressionType getOutputType(InputBindingInspector inspector)
+          @Override
+          public Expr visit(Shuttle shuttle)
+          {
+            List<Expr> newArgs = args.stream().map(x -> x.visit(shuttle)).collect(Collectors.toList());
+            return shuttle.visit(new CastJsonValueExpr(newArgs));
+          }
+
+          @Nullable
+          @Override
+          public ExpressionType getOutputType(InputBindingInspector inspector)
+          {
+            return castTo;
+          }
+        }
+        return new CastJsonValueExpr(args);
+      } else {
+        class JsonValueExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
         {
-          // we cannot infer the output type (well we could say it is 'STRING' right now because is all we support...
-          return null;
+          public JsonValueExpr(List<Expr> args)
```

Comment:
Seems like the logic is the same between `get_path` and the no-cast version of `json_value`. Can we share the code too? Maybe use the same static class for both?

Reply:
yeah, `get_path` is sort of a relic from my earliest prototypes, and used to look a fair bit different (and only accepted 'jq' style paths), but does look the same now. Will consolidate it. Maybe we need an easy way to alias native expressions like we have for SQL expressions... I have a similar situation going on with `json_object` which is a wrapper around the earlier prototype `struct` function.

Reply:
Since we haven't done a release yet with this functionality, would it make sense to remove some of these functions? Are they going to be "born obsolete"?

Reply:
Yeah that's a good point. I think i'll remove them. The only one I'm a bit partial to keeping is `struct` since longer term I sometimes imagine replacing `COMPLEX<json>` with `STRUCT` and `VARIANT` types so that the engine can be integrated a bit tighter with nested data columns. But, I can add it back whenever we get to that point, so will remove it too for now.